src/main/java/httpserver/HTTPRequest.java

```diff
@@ -25,509 +25,502 @@
+        }
+
+        try {
+            parseRequest();
+            HTTPResponse response = new HTTPResponse(this);
+            determineHandler().handle(this, response);
+            response.respond();
+        } catch (IOException | HTTPException e) {
+            e.printStackTrace();
+        }
+    }
 
-      String[] items = line.split(": ");
 
-      if (items.length == 1) {
-        throw new HTTPException("No key value pair in \n\t" + line);
-      }
+    /**
+     * Kicks off the request's parsing. Called inside constructor.
+     *
+     * @throws IOException      When an InputStream can't be retreived from the
+     *                          socket.
+     * @throws SocketException  When the client breaks early. This is a browser
+     *                          issue, and not a server issue, but it gets thrown
+     *                          upstream because it can't be dealt with until it
+     *                          gets to the HTTPServer.
+     * @throws HTTPException    When headers aren't in key/value pairs separated
+     *                          by ": ".
+     *
+     * @see HTTPServer
+     */
+    public void parseRequest() throws IOException, SocketException, HTTPException {
+        // Used to read in from the socket
+        BufferedReader input = new BufferedReader(
+                new InputStreamReader(getConnection().getInputStream()));
+
+        StringBuilder requestBuilder = new StringBuilder();
+
+        /*  The HTTP spec (Section 4.1) says that a blank first line should be
+            ignored, and that the next line SHOULD have the request line. To be
+            extra sure, all initial blank lines are discarded.
+            */
+        String firstLine = input.readLine();
+        if (firstLine == null) {
+            throw new HTTPException("Input is returning nulls...");
+        }
+
+        while (firstLine.isEmpty()) {
+            firstLine = input.readLine();
+        }
+
+        // start with the first non-empty line.
+        setRequestLine(firstLine);
+        requestBuilder.append(getRequestLine());
+        requestBuilder.append("\n");
+
+        /*  Every line after the first, but before an empty line is a header,
+            which is a key/value pair.
+
+            The key is before the ": ", the value, after
+
+            TODO: parse this to spec. Spec says it's cool to have any number of
+            whitespace characters following the colon, and the values
+            can be spread accross multiple lines provided each following line
+            starts with a whitespace character.
+
+            For more information, see issue 12 and RFC 2616#4.2.
+            Issue 12: https://github.com/dkuntz2/java-httpserver/issues/12
+            RFC 2616#4.2: http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+            */
+        for (String line = input.readLine(); line != null && !line.isEmpty(); line = input.readLine()) {
```

Comment:
Does `BufferedReader#readLine` return null, or just say it's empty? (Do we need the `line != null` check?)